python/hail/utils/misc.py

```diff
@@ -260,12 +260,53 @@ def get_nice_field_error(obj, item):
 def check_collisions(fields, name, indices):
     from hail.expr.expressions import ExpressionException
     if name in fields and not fields[name]._indices == indices:
-        msg = 'name collision with field indexed by {}: {}'.format(list(fields[name]._indices.axes), name)
+        msg = "name collision with field indexed by {}: '{}'".format(list(fields[name]._indices.axes), name)
         error('Analysis exception: {}'.format(msg))
         raise ExpressionException(msg)
 
 def check_field_uniqueness(fields):
     for k, v in Counter(fields).items():
         if v > 1:
             from hail.expr.expressions import ExpressionException
-            raise ExpressionException("selection would produce duplicate field {}".format(repr(k)))
+            raise ExpressionException("selection would produce duplicate field '{}'".format(repr(k)))
+
+def check_keys(name, indices):
+    from hail.expr.expressions import ExpressionException
+    if name in set(indices.key):
+        msg = "cannot overwrite key field '{}' with annotate or select; use key_by to modify keys.".format(name)
+        error('Analysis exception: {}'.format(msg))
+        raise ExpressionException(msg)
+
+def get_select_exprs(caller, exprs, named_exprs, indices, protect_keys=True):
+    from hail.expr.expressions import to_expr, ExpressionException, TopLevelReference, Select
+    exprs = [to_expr(e) if not isinstance(e, str) else indices.source[e] for e in exprs]
+    named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
+    assignments = OrderedDict()
+
+    for e in exprs:
+        if not e._indices == indices:
+            raise ExpressionException("method '{}' parameter 'exprs' expects {}-indexed fields,"
+                                  " found indices {}".format(caller, list(indices.axes), list(e._indices.axes)))
+        if e._ast.search(lambda ast: not isinstance(ast, TopLevelReference) and not isinstance(ast, Select)):
```

Comment:
I don't like this. I know it's used elsewhere, but ick.

Maybe OK for now, but how should we fix? Store a dict of all nested fields in the object? That might do it.

Reply:
you mean like _inverse_fields, but nested? I could do that.

Reply:
would that maybe get super large, though?

Reply:
you're right. How about we just abstract this to a function in the AST module, `is_nested_reference`?

Reply:
ok. let me know if what I did is too silly.